File: remote-control/agent/internal/upgrade/manifest.go
package upgrade

import (
	"encoding/json"
	"fmt"
	"net/url"
	"strings"
)

type Manifest struct {
	Version                  string `json:"version"`
	DownloadURL              string `json:"downloadUrl"`
	SHA256                   string `json:"sha256"`
	SizeBytes                int64  `json:"sizeBytes"`
	MinimumCompatibleVersion string `json:"minimumCompatibleVersion,omitempty"`
}

func ParseManifest(data []byte) (Manifest, error) {
	var manifest Manifest
	if err := json.Unmarshal(data, &manifest); err != nil {
		return Manifest{}, fmt.Errorf("parse manifest: %w", err)
	}
	if err := ValidateManifest(manifest); err != nil {
		return Manifest{}, err
	}
	return manifest, nil
}

func ValidateManifest(manifest Manifest) error {
	if strings.TrimSpace(manifest.Version) == "" {
		return fmt.Errorf("manifest version is required")
	}
	if strings.TrimSpace(manifest.DownloadURL) == "" {
		return fmt.Errorf("manifest downloadUrl is required")
	}
	parsed, err := url.Parse(manifest.DownloadURL)
	if err != nil || parsed.Scheme == "" || parsed.Host == "" {
		return fmt.Errorf("manifest downloadUrl is invalid")
	}
	if strings.TrimSpace(manifest.SHA256) == "" {
		return fmt.Errorf("manifest sha256 is required")
	}
	if manifest.SizeBytes <= 0 {
		return fmt.Errorf("manifest sizeBytes must be positive")
	}
	return nil
}

File: remote-control/server/src/upgrades/manifest.js
function currentApprovedManifest() {
  const version = process.env.UPGRADE_VERSION || '';
  const downloadUrl = process.env.UPGRADE_DOWNLOAD_URL || '';
  const sha256 = process.env.UPGRADE_SHA256 || '';
  const sizeBytes = Number(process.env.UPGRADE_SIZE_BYTES || 0);
  const minimumCompatibleVersion = process.env.UPGRADE_MIN_COMPATIBLE_VERSION || '';

  if (!version || !downloadUrl || !sha256 || !sizeBytes) {
    return null;
  }

  return {
    version,
    downloadUrl,
    sha256,
    sizeBytes,
    ...(minimumCompatibleVersion ? { minimumCompatibleVersion } : {}),
  };
}

function validateManifest(manifest) {
  if (!manifest) return false;
  try {
    const parsed = new URL(manifest.downloadUrl);
    return Boolean(
      manifest.version &&
      parsed.protocol.startsWith('http') &&
      manifest.sha256 &&
      Number(manifest.sizeBytes) > 0
    );
  } catch {
    return false;
  }
}

module.exports = {
  currentApprovedManifest,
  validateManifest,
};
